sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala

```diff
@@ -556,11 +557,56 @@ object Range {
     val output = StructType(StructField("id", LongType, nullable = false) :: Nil).toAttributes
     new Range(start, end, step, numSlices, output, isStreaming)
   }
+
   def apply(start: Long, end: Long, step: Long, numSlices: Int): Range = {
     Range(start, end, step, Some(numSlices))
   }
+
+  private def castAndEval[T](expression: Expression, dataType: DataType): T = {
+    TypeCoercion.implicitCast(expression, dataType)
```

Comment:
shall we use `AnsiTypeCoercion` instead?

Reply:
What's the behavior before this PR? We probably don't have to follow the type coercion rules and only allow `IntegralType`.

Reply:
Before this PR the arguments to `range` will be implicitly casted. For example
```sql
select * from range(6 + cos(3));  -- double -> long
select * from range('1');  -- string -> long
```
To be backward compatible we should still support these cases, but they need to respect the ansi mode when enabled.